### PR TITLE
Rdf2alma

### PR DESCRIPTION
**Pull request by jimfhahn**
Hi Jeremy,

This request includes changes to the Penn DAG to send RDF from Sinopia to Alma. I have only made changes to the Penn DAG for this PR. My thinking was that once this is worked out, I'll add the remainder of the focus group in the next PR. 

Tests for the new code are included. I somehow got the poetry lock out of version with the test environment while adding something. Let me know what suggestions you have for that and any other parts of the PR. 

-Jim

### CODE REVIEW

**Comment by jimfhahn** [comment]
> @jimfhahn we just merged into main a fix for the poetry issues if you want to rebase. I'll then approve and merge this PR. Thanks!

@jermnelson good to go! 👍

File: ils_middleware/tasks/alma/post_bfwork.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+            "Content-Type": "application/xml; charset=utf-8",
+            "Accept": "application/xml",
+            "x-api-key": alma_api_key,
+        },
+        data=data,
+    )
+    logger.debug(f"alma result: {alma_result.status_code}\n{alma_result.text}")
+    result = alma_result.content
+    status = alma_result.status_code
+    xml_response = ET.fromstring(result)
```
Comment (jermnelson) [inline ils_middleware/tasks/alma/post_bfwork.py:55]: Maybe create a new function that takes the result and does the xslt transform and returns the `mms_id` here would help drop the complexity of the function

i.e.
```
mmis_id = parseMMSId(result)
.
.
.
def parseMMSId(result):
    xml_response = ET.fromstring(result)
    xslt = ET.parse("ils_middleware/tasks/alma/xslt/put_mms_id.xsl")
    transform = ET.XSLT(xslt)
    result_tree = transform(xml_response)
    put_mms_id_str = str(result_tree)
    logger.debug(f"put_mms_id_str: {put_mms_id_str}")
    mms_id = xml_response.xpath("//mms_id/text()")
    logger.debug(f"mms_id: {mms_id}")
    return mms_id
```

File: ils_middleware/tasks/alma/post_bfwork.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    transform = ET.XSLT(xslt)
+    result_tree = transform(xml_response)
+    put_mms_id_str = str(result_tree)
+    logger.debug(f"put_mms_id_str: {put_mms_id_str}")
+    mms_id = xml_response.xpath("//mms_id/text()")
+    logger.debug(f"mms_id: {mms_id}")
+    if status == 200:
+        task_instance.xcom_push(key=instance_uri, value=mms_id)
+    elif status == 400:
+        alma_update_uri = (
```
Comment (jermnelson) [inline ils_middleware/tasks/alma/post_bfwork.py:67]: Another place where calling a function to handle the PUT would help reduce code complexity.

i.e.
```
def putWorkToAlam(task_instance, put_mms_id_str, alma_api_key):
      alma_update_uri = (
          uri_region
            + "/almaws/v1/bibs/"
            + put_mms_id_str
            + "?normalization=&validate=false&override_warning=true"
            + "&override_lock=true&stale_version_check=false&cataloger_level=&check_match=false"
            + "&apikey="
            + alma_api_key
        )
        put_update = requests.put(
            alma_update_uri,
            headers={
                "Content-Type": "application/xml; charset=UTF-8",
                "Accept": "application/xml",
                "x-api-key": alma_api_key,
            },
            data=data,
        )
              logger.debug(f"put update: {put_update.status_code}\n{put_update.text}")
        put_update_status = put_update.status_code
        result = put_update.content
        xml_response = ET.fromstring(result)
        put_mms_id = xml_response.xpath("//mms_id/text()")
        if put_update_status == 200:
            task_instance.xcom_push(key=instance_uri, value=put_mms_id)
        if put_update_status != 200:
            raise Exception(f"Unexpected status code: {status}")

File: docker-compose.yaml [hunk 3]
```diff
@@ -63,7 +63,7 @@ x-airflow-common:
     AIRFLOW_VAR_MARC_S3_BUCKET: 'sinopia-marc-development'
     AIRFLOW_VAR_RDF2MARC_LAMBDA: 'sinopia-rdf2marc-development'
     AIRFLOW_VAR_SINOPIA_ENV: ${AIRFLOW_VAR_SINOPIA_ENV:-dev}
-    AIRFLOW_VAR_SQS_URL: 'http://localstack:4566/000000000000/'
+    #AIRFLOW_VAR_SQS_URL: 'http://localstack:4566/000000000000/'
```
Comment (jermnelson) [inline docker-compose.yaml:66]: Have you tried testing with this variable with the update to Airflow?

File: ils_middleware/tasks/alma/post_bfinstance.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    transform = ET.XSLT(xslt)
+    result_tree = transform(xml_response)
+    put_mms_id_str = str(result_tree)
+    logger.debug(f"put_mms_id_str: {put_mms_id_str}")
+    mms_id = xml_response.xpath("//mms_id/text()")
+    logger.debug(f"mms_id: {mms_id}")
+    if status == 200:
+        task_instance.xcom_push(key=instance_uri, value=mms_id)
+    elif status == 400:
+        alma_update_uri = (
```
Comment (jermnelson) [inline ils_middleware/tasks/alma/post_bfinstance.py:67]: See comment below with an example of moving this PUT functionality to a separate function to help reduce code complexity

File: ils_middleware/tasks/amazon/alma_instance_s3.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+            reader = MARCReader(marc_file)
+            for record in reader:
+                instance_field = record.get_fields("884")
+                instance_uri = instance_field[0].get_subfields("k")[0]
+                logger.info(f"Instance URI: {instance_uri}")
+            h = Graph()
+            h.parse(instance_uri)
+            # declare namespaces
+            h.bind("xmlns", "http://www.w3.org/2000/xmlns/")
+            h.bind("xsd", "http://www.w3.org/2001/XMLSchema#")
```
Comment (jermnelson) [inline ils_middleware/tasks/amazon/alma_instance_s3.py:53]: Not required; `xsd` is one of the default namespaces already bound to a rdflib Graph.

File: ils_middleware/tasks/amazon/alma_instance_s3.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+            h.bind("rdf", "http://www.w3.org/1999/02/22-rdf-syntax-ns#")
```
Comment (jermnelson) [inline ils_middleware/tasks/amazon/alma_instance_s3.py:54]: Not required; `rdf` is one of the default namespaces already bound to a rdflib Graph.

File: ils_middleware/tasks/amazon/alma_instance_s3.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+            h.bind("rdfs", "http://www.w3.org/2000/01/rdf-schema#")
```
Comment (jermnelson) [inline ils_middleware/tasks/amazon/alma_instance_s3.py:55]: Not required; `rdfs` is one of the default namespaces already bound to a rdflib Graph.

File: ils_middleware/tasks/amazon/alma_instance_s3.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+            h.bind("bf", "http://id.loc.gov/ontologies/bibframe/")
+            h.bind("bflc", "http://id.loc.gov/ontologies/bflc/")
+            h.bind("madsrdf", "http://www.loc.gov/mads/rdf/v1#")
+            h.bind("sinopia", "http://sinopia.io/vocabulary/")
+            h.bind("sinopiabf", "http://sinopia.io/vocabulary/bf/")
+            h.bind("rdau", "http://rdaregistry.info/Elements/u/")
+            h.bind("owl", "http://www.w3.org/2002/07/owl#")
```
Comment (jermnelson) [inline ils_middleware/tasks/amazon/alma_instance_s3.py:62]: Not required; `owl` is one of the default namespaces already bound to a rdflib Graph.

File: ils_middleware/tasks/amazon/alma_work_s3.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+        marc_record_from_temp_file(instance_id, temp_file)
+        with open(temp_file, "rb") as marc_file:
+            reader = MARCReader(marc_file)
+            for record in reader:
+                work_field = record.get_fields("758")
+                work_uri = work_field[0].get_subfields("0")[0]
+                logger.info(f"Work URI: {work_uri}")
+            g = Graph()
+            g.parse(work_uri)
+            g.bind("xmlns", "http://www.w3.org/2000/xmlns/")
```
Comment (jermnelson) [inline ils_middleware/tasks/amazon/alma_work_s3.py:51]: See suggestions above for using a `for` loop on a common `alma_namespaces` variable.

**Review by jermnelson — CHANGES_REQUESTED** [review]
Overall, looks good. Some comments for reducing code complexity to pass code climate (see https://codeclimate.com/github/LD4P/ils-middleware/pull/170) check. 

Also, a couple of suggestions on DRY namespace binding for rdf graphs.

File: docker-compose.yaml (hunk 3 above)
Comment (jimfhahn) [reply docker-compose.yaml:66]: working again with the airflow update

**Review by jermnelson — APPROVED** [review]

**Pull request merged**